contracts/schemes/ArcUpgradeScheme.sol

```diff
@@ -0,0 +1,200 @@
+        address[] memory contractsToUpgrade = proposal.contractsToUpgrade;
+        for (uint256 i = 0; i < contractsToUpgrade.length; i++) {
+            bytes32 contractNameBytes = proposal.contractsNames[i];
+            string memory contractName = bytes32ToStr(contractNameBytes);
+            address updatedImp = ImplementationProvider(
+                package.getContract(proposal.packageVersion)
+            ).getImplementation(contractName);
+
+            Controller controller = Controller(avatar.owner());
+            controller.genericCall(
+                contractsToUpgrade[i],
+                abi.encodeWithSignature("upgradeTo(address)", updatedImp),
+                0
+            );
+        }
+
+        delete organizationProposals[_proposalId];
+        emit ProposalDeleted(address(avatar), _proposalId);
+        emit ProposalExecuted(address(avatar), _proposalId);
+    }
+
+    /**
+    * @dev propose upgrade contracts Arc version
+    *      The function trigger NewUpgradeProposal event
+    * @param _packageVersion - the new Arc version to use for the contracts
+    * @param _contractsNames - names of contracts which needs to be upgraded
+    * @param _contractsToUpgrade - addresses of contracts which needs to be upgraded
+    * @param _descriptionHash - proposal description hash
+    * @return an id which represents the proposal
+    */
+    function proposeUpgrade(
+        uint64[3] memory _packageVersion,
+        bytes32[] memory _contractsNames,
+        address[] memory _contractsToUpgrade,
+        string memory _descriptionHash)
+    public
+    returns(bytes32)
+    {
+        require(package.hasVersion(_packageVersion), "Specified version doesn't exist in the Package");
+        for (uint256 i = 0; i < _contractsToUpgrade.length; i++) {
+            require(
+                ImplementationProvider(
+                    package.getContract(_packageVersion)
+                ).getImplementation(bytes32ToStr(_contractsNames[i])) != address(0),
+                "Contract name does not exist in ArcHive package"
+            );
+        }
+
+        bytes32 proposalId = votingMachine.propose(2, voteParams, msg.sender, address(avatar));
+
+        organizationProposals[proposalId] = UpgradeProposal({
+            packageVersion: _packageVersion,
+            contractsNames: _contractsNames,
+            contractsToUpgrade: _contractsToUpgrade,
+            exist: true,
+            passed: false
+        });
+        proposalsInfo[address(votingMachine)][proposalId] = ProposalInfo({
+            blockNumber:block.number,
+            avatar:avatar
+        });
+        emit NewUpgradeProposal(
+            address(avatar),
+            proposalId,
+            _packageVersion,
+            _contractsNames,
+            _contractsToUpgrade,
+            _descriptionHash
+        );
+        return proposalId;
+    }
+
+    function bytes32ToStr(bytes32 x) private pure returns (string memory) {
```

Comment:
this bit complex util function is also at https://github.com/daostack/arc/blob/arc-factory/contracts/utils/DAOFactory.sol#L234 .
how about make it in a lib and use it in both places ?

Reply:
moved to a lib